### ISSUE TITLE
Gzip based on response content type

### ISSUE THREAD

**Issue by sajal** [NYTimes/gziphandler#40]
I see now `GzipResponseWriter` can [decide](https://github.com/NYTimes/gziphandler/blob/9d655ed18f36edb11d06469a0aefee31458d96a4/gzip.go#L107) if it wants to gzip the response based on response size. 

I suggest also filtering on content-types. The types could be user configured or hardcoded to [sane defaults](https://github.com/h5bp/server-configs-nginx/blob/cb3dc0554eb54ccfab34bfcf32988dfce1872ec2/nginx.conf#L91)

	if _, ok := w.Header()[contentType]; !ok {
		// It infer it from the uncompressed body.
		w.Header().Set(contentType, http.DetectContentType(b))
	}
	//Use the content-type information to decide between `w.startGzip()` and `w.startNoGzip()`

**Comment by flimzy** [comment]
This sounds like a nice idea, but the combinatorial explosion of config options might be getting out of hand.

**Comment by sajal** [comment]
How about a config struct with sane defaults?

**Comment by adammck** [comment]
I like config structs, but I don't think this functionality belongs in this middleware. I'm not strongly against it, and the implementation would be trivial, but I'd like to hear some concrete use-cases first. I rarely encounter endpoints which may or may not want their responses compressed.

**Comment by sajal** [comment]
For API servers, sending content uncompressed is not needed. For this use-case the current behaviour is perfect. This is the primary way I too use this middleware.

But if I am writing a server that serves full website (images, videos, binary, etc) then gziping the payload is often redundant. Practically no full-feature webserver would compress jpg, png, mp4, etc since these formats dont have enough gain. This is especially true for videos... running on-the-fly gzip on 1 GB video file to save few KBs is quite wasteful. In [this case](https://github.com/turbobytes/dummyorigin/blob/14e1267d06ba07a8ab15a3ffd49d17f936cd0377/main.go#L179-L205), I had to write my own filter based on file-extension to enable/disable compression, but I would rather do it based off response Content-Type.

**Comment by romainmenke** [comment]
On config / options : [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)

Personally I really like this approach.

**Comment by adammck** [comment]
@romainmenke: [me, too](https://github.com/NYTimes/gziphandler/issues/43) :smile:

**Comment by ConradIrwin** [comment]
I'd like this — we proxy assets through go right now (which we arguably shouldn't), and it'd be nice to compress the javascript files but not the pngs.

**Comment by holygeek** [comment]
Another point to take into consideration is to not gzip an already gzipped content, something like this:

```
$ git diff
diff --git gzip.go gzip.go
index c47039f..36ccd1b 100644
--- gzip.go
+++ gzip.go
@@ -102,10 +102,11 @@ func (w *GzipResponseWriter) Write(b []byte) (int, error) {
        // On the first write, w.buf changes from nil to a valid slice
        w.buf = append(w.buf, b...)
 
+       _, alreadyEncoded := w.Header()[contentEncoding]
        // If the global writes are bigger than the minSize and we're about to write
        // a response containing a content type we want to handle, enable
        // compression.
-       if len(w.buf) >= w.minSize && handleContentType(w.contentTypes, w) {
+       if !alreadyEncoded && len(w.buf) >= w.minSize && handleContentType(w.contentTypes, w) {
                err := w.startGzip()
                if err != nil {
                        return 0, err
```

I'm currently using `gziphandler` in a server that serves content either "locally" from my server, or reverse proxied off from other servers. Currently I'm using a stack of http handlers that includes gziphandler, and for all requests that are to be proxied I do `r.Header.Del("Accept-Encoding")` which means the content will be gzipped "locally" by gziphandler. To avoid this (and to reduce CPU usage on my server) I would have to create two stacks of http handlers -one with gziphandler and another without so that all the gzipping is done at the proxied servers and my server would then just pass the response through without modification. If gziphalder had the above code then I don't have to create two stacks for http handlers, which could simplify my code.

**Comment by jprobinson** [comment]
You all win :grin: This has been resolved with #55

**Comment by flimzy** [comment]
That PR resolves the issue of not recompressing responses with the content-encoding header, but this issue is about not compressing responses based on the content-type header.

**Comment by tmthrgd** [comment]
This was fixed with the merging of #51.
